### ISSUE TITLE
LLVM Intrinsics for Trig. Functions

### ISSUE THREAD

**Issue by sclel016** [JuliaLang/julia#41763]
Hi,

I noticed that Julia implements `sin` and `cos` based on polynomials from `openlibm` in [trig.jl](https://github.com/JuliaLang/julia/blob/f711f0a9076b9f11f9034985086d73bc285c2ca6/base/special/trig.jl). Since Julia relies so heavily on LLVM, I was surprised that the trigonometric intrinsics were not being leveraged. 

I did a couple quick benchmarks to verify performance. The first is on `N` shuffled values in the range `[0,pi/4]` to test for accuracy.  There was at most a single epsilon of error with LLVM being slightly faster.

```
using BenchmarkTools
N = convert(Int,1E6);
x1 = collect(range(0, pi/4, length=N));
x = shuffle(x1);
y_jl = similar(x);
@btime @. y_jl = sin(x);
  2.600 ms (2 allocations: 32 bytes)

y_llvm = similar(x);
llvm_sin(x) = ccall("llvm.sin.f64",llvmcall,Float64,(Float64,),x);
@btime @. y_llvm = llvm_sin(x);
  2.332 ms (2 allocations: 32 bytes)

maximum(abs.(y_llvm - y_jl))
1.1102230246251565e-16
```

Next I tested random floats, this time for speed while messing with the input range. LLVM is still slightly faster while maintaining at most a single eps of error.

```
using BenchmarkTools
N = convert(Int,1E6);
x = rand(Float64,N);
y_jl = similar(x);
@btime @. y_jl = sin(x);
    4.575 ms (2 allocations: 32 bytes)

y_llvm = similar(x);
llvm_sin(x) = ccall("llvm.sin.f64",llvmcall,Float64,(Float64,),x);
@btime @. y_llvm = llvm_sin(x);
    4.287 ms (2 allocations: 32 bytes)

maximum(abs.(y_llvm - y_jl))
1.1102230246251565e-16
```

In short, why isn't Julia using the LLVM trig. intrinsics when possible? Is there an accuracy or performance cost I missed?

**Comment by Keno** [comment]
LLVM's trig intrinsics will forward to your platform's libm, which can be of dubious quality. The julia implementation ensures that results are both reproducable and high quality on all platforms. There are some benefits to the LLVM intrinsics, namely that they can fold late and vectorize, but that is better solved by adding the ability for LLVM to do that generically for all julia functions rather than a select few special cases.

**Comment by sclel016** [comment]
I would have thought another benefit of using LLVM intrinsics would be the HW support. On x86, it seems that LLVM uses `fsin` based on the [source](https://github.com/llvm/llvm-project/blob/6da9241aabe12c07fb905e5b3d2e4e16ef7d4296/libc/src/math/x86_64/sin.cpp) while the [generic method](https://github.com/llvm/llvm-project/blob/3a41ff4883fe8b9e34a4f30aa9eecaf2ecb2ef44/libc/src/math/generic/sincosf_utils.h) uses a polynomial approximation that doesn't look as clean as the Julia implementation. I take it there's no concern with the accuracy of `fsin` but a general distrust of the generic approximations?

**Comment by Keno** [comment]
As I said, it forwards to system libm. LLVM has an implementation, but it's unlikely that that's the one you're using. Also `fsin` is frowned upon on modern CPUs because it had terrible performance, since it's part of the legacy x87 support.

**Comment by Keno** [comment]
Oh also, `fsin` is bad at range reduction and loses a bunch of accuracy there.

**Comment by sclel016** [comment]
Thank for the insight. My system is indeed using glibc, but I guess the point in general still stands. I'll go ahead and close the issue, thank!

**Comment by oscardssmith** [comment]
To be more specific, Julia's implementation maintains <=.8 ULPs over the entire range (including things like `sin(1.3e54)`). Although it can seem like anything between 1 and .5 is in the same boat, having higher accuracy makes it easier to keep more complicated things (like `sin(::Complex)`) accurate enough.
